doc/toolchains.md

```diff
@@ -0,0 +1,80 @@
+cross-compilation). Note that, independently of the compiler finally made
+available by `alr` in the environment, the crate project file still must define
+an appropriate `Target` attribute for the desired compiler. At the moment, Alire
+does not examine project file contents to identify necessary compilers, and
+relies only on regular `depends-on` dependencies.
+
+There are two sides to toolchain use by `alr`. On the one hand, a solution may
+or may not have dependencies on GNAT compilers. On the other hand, the user may
+or may have not selected a default toolchain for use via `alr toolchain
+--select`. The interaction between these two features is explained next.
+
+Before going into the details, there are two kind of dependencies on GNAT
+compilers: generic dependencies on the `gnat` crate, which apply to every 
+compiler, and dependencies on a precise native or cross-compiler, e.g.,
+`gnat_native` or `gnat_riscv_elf`.
+
+## Identifying available compilers
+
+Available compilers can be listed with `alr search --full --external-detect
+gnat_`. They will also be shown by the selection assistant, `alr toolchain
+--select`.
+
+## Specifying dependencies on GNAT compilers
+
+From the point of view of a user preparing a release for publication, these
+are the considerations to bear in mind:
+
+- Do not use any dependency on a compiler if you do not have a reason to do so.
+- Use dependencies on `gnat` to specify known compiler version restrictions.
+- Use dependencies on a precise gnat cross-compiler (e.g., `gnat_riscv_elf`)
+  when your crate is purposely only available for such a target.
+- There is no reason to specify a dependency on the native compiler
+  (`gnat_native`) as that would unnecessarily limit the availability of a
+  library crate that might be useful to dependent cross-target crates.
+
+## Interactions with a selected toolchain
+
+From the point of view of a user wanting to compile some release, there will be
+an interaction between the solution dependency on compilers and a selected
+default compiler. 
+
+The simplest and most usual scenario for native compilation is as follows:
+
+- Solution without dependencies on `gnat`: 
+   - If a default compiler has been configured: the compiler will be included in the
+     PATH provided by `alr printenv` and used for the build.
+   - No compiler has been configured: `alr` will attempt to use whatever toolchain is
+     available in the user's environment prior to running `alr`. 
+      - Note that compilers deployed with `alr toolchain --install` but not
+        selected as the default will not be used in this case.
+
+For crates with dependencies on GNAT compilers, the following two cases simultaneously apply:
+
+- Solution with generic dependencies on `gnat`: The solver will provide a
+  concrete GNAT to satisfy this dependency, applying the following
+  prioritization:
+   1. A target-specific compiler that is already a dependency.
```

Comment:
This is how the solver tries to find a release for `gnat`. Do you mean that the solver should not bring in compilers that aren't already installed? (At the moment, if a crate requests a compiler and none is available, it will pick one in the index and auto-install it).

Reply:
> Do you mean that the solver should not bring in compilers that aren't already installed? (At the moment, if a crate requests a compiler and none is available, it will pick one in the index and auto-install it).

Not if the user didn't specify a default, and there is no explicit dependency on `gnat_native`, `gnat_arm_elf`, etc.

If the user decided not to have a default native compiler, which I would say I hope not to be a very common case, then for me it means he/she is in charge of providing the compiler.

Reply:
OK, I understand. So `gnat` dependencies only rely on locally available compilers. Should be easy to fix.

Reply:
This is implemented and checked in a new test in the last commit (8cca83d98e0185addda06a55c97fccdc19aa22e6).

As of now, and correcting my last message, to avoid any surprises for the user, an external compiler will be preferred unless the user selects its preference, or the crate specifies a targeted gnat (`gnat_something`).

BTW while installed compilers are for the whole config, we can make the selected compiler optionally a local config option, so different working crates may select a different compiler among the installed ones. This can be a follow-up PR.